### DISCUSSION TITLE
Shouldn't {{< include >}} follow the project's `execute-dir` option?

### DISCUSSION THREAD

**Discussion by edalfon** [quarto-dev/quarto-cli#3784] [Q&A]
Hi,

We use [includes](https://quarto.org/docs/authoring/includes.html) to re-use content across documents in a Quarto website project. 

In this project, we set the [option `execute-dir: project`](https://quarto.org/docs/reference/projects/websites.html) to keep the working directory for computations pinned to the root directory of the project. That way, for example, all documents can refer to `./data/fancy_data`, regardless of which subfolder they are placed in. For us this is important because we organize the content of the project in a bunch of subfolders that can go nested to different depth. Without the `execute-dir: project` the project would simply be much harder to mantain and keep track of the paths we should use in each document.

We expected `{{< include >}}` to follow the same project's `execute-dir` option, but it doesn't seem so. Instead, it always resolves to the same folder of the file you are including content in. So if you are in a subfolder and whant to include some content in the root of the project, you cannot do this `{{< include _codebit.qmd>}}`, but you have to do something like this `{{< include ../../../_codebit.qmd>}}`. And the `../../...` part would depend on where the document lives and if you decide to reorganize the structure of the documents, you would need to update that as well. So, it is a bit of a mess.

We couldn't find a way around this or any other option similar to `execute-dir`, but for the whole working directory of the project and not only for computations.

Wouldn't it be better to have the includes directive to follow the `execute-dir` option?, is there currently any way to do it?

**Comment by mcanouil** [comment]
> We expected {{< include >}} to follow the same project's execute-dir option

Why did you expect this? execute options are for computation only, it has nothing to do with all files/resources are located by their path. Files/resources are relatives to the document using them.

> Wouldn't it be better to have the includes directive to follow the execute-dir option?

This would require all Quarto files to have unique names.
I do not believe what you are asking should be implemented at all, because contrary to what you might think, it would be a nightmare to navigate through the files if when looking at the include shortcode or any options that include resources, you have to think twice to find out to what it is relative to.
Also, this would lead to inconsistent behaviour, if options use path relative to the document, while include use path relative to project root, etc.

**Reply by edalfon** [reply]
Hi!, Thanks for your answer. A couple of further questions just to understand your point of view.

> This would require all Quarto files to have unique names. 

Why would they have to be unique?

> you have to think twice to find out to what it is relative to. 

Yeah, exactly as you currently have to think twice when you use a path in any computation chunk. You have to think to what will it be relative. And it will be relative either to the file, or the project (as configured via the `exec-dir` option). Why that would be a problem for includes, and it is not a problem for computations?

**Reply by mcanouil** [reply]
> Why would they have to be unique?

I spoke too quickly, I realised your "include example" assumed the file was at the root directory.


> Yeah, exactly as you currently have to think twice when you use a path in any computation chunk. You have to think to what will it be relative. And it will be relative either to the file, or the project (as configured via the exec-dir option). Why that would be a problem for includes, and it is not a problem for computations?

You need the working directory at the root of the project when you have a shared setup for computation, such as conda, renv, etc. So this option is absolutely required/needed.
This option only ensure the session (of whatever language) as the set working directory either to the Quarto file or the root directory.
For resources inside a Quarto document, I think it's cumbersome to see `./file.qmd` and having to think `.` is not the current directory.
By the way, this is how knitr child works as well for R Markdown.


To note, I am not saying it's a bad idea to implement an option for the behaviour you want, I am saying it should absolutely not be tied up to the `execute-dir` option.

**Reply by dragonstyle** [reply]
In Quarto projects, an include path that begins with a leading slash will be interpreted as project relative, meaning that you should be able to use an include such as:

```
{{< include /_codebit.qmd >}}
```

to include from the project root no matter the subdirectory of the file itself.

HTH!

**Reply by edalfon** [reply]
thx for the clarifications. 

> To note, I am not saying it's a bad idea to implement an option for the behaviour you want, I am saying it should absolutely not be tied up to the `execute-dir` option.

Yeah, I agree. There should probably be two independent options.

**Reply by edalfon** [reply]
> In Quarto projects, an include path that begins with a leading slash will be interpreted as project relative, meaning that you should be able to use an include such as:
> 
> ```
> {{< include /_codebit.qmd >}}
> ```
> 
> to include from the project root no matter the subdirectory of the file itself.
> 
> HTH!

Awesome!, thank you very much. Will try that ... (we didn't try that one, as we thought that would just point to the root of the drive/system, sort of like c: in windows)

**Reply by mcanouil** [reply]
@dragonstyle Maybe this "trick" should be mentioned/highlighted in the documentation.

https://quarto.org/docs/authoring/includes.html

**Reply by dragonstyle** [reply]
Yeah I was thinking it might make sense to document more generally where 'project paths' can be used (they work in many other places of Quarto as well). I can make a note in the include docs as well.

**Reply by davidpomerenke** [reply]
> In Quarto projects, an include path that begins with a leading slash will be interpreted as project relative, meaning that you should be able to use an include such as:
> 
> ```
> {{< include /_codebit.qmd >}}
> ```
> 
> to include from the project root no matter the subdirectory of the file itself.
> 
> HTH!

This is quite handy, but it does not seem to work for embeds. Maybe it makes sense to implement the same mechanism there?

My use case is that I would like to use nested includes (which works!), some of which also have embeds (depending on how I specify the path, these work either only when rendering the composite document, or only when rendering the (sub-)include). My workaround for rendering at both levels is using absolute paths, which is totally okay for my project, but maybe not for larger collaborative projects. It is of course not a given that it should be possible to render both the (sub-)include and the composite document, but I find it quite handy to get faster renders for the parts that I am working on.

**Reply by mcanouil** [reply]
@davidpomerenke I opened a dedicated issue #6147
